src/ddmd/intrange.d

```diff
@@ -494,4 +524,233 @@ struct IntRange
+            auto maxAndPos = maxAnd(l, R(SignExtendedNumber(0), r.imax));
+
+            auto min = minAndNeg < minAndPos ? minAndNeg : minAndPos;
+            auto max = maxAndNeg > maxAndNeg ? maxAndNeg : maxAndPos;
+
+            return R(min, max);
+        }
+    }
+
+    // Credits to Timon Gehr for the algorithms for &, |
+    // https://github.com/tgehr/d-compiler/blob/master/vrange.d
+    R opBinary(string op:"|")(R rhs) const
+    {
+        // unsigned or identical sign bits:
+        if ((imin.value ^ imax.value) >= 0 && (rhs.imin.value ^ rhs.imax.value) >= 0)
+        {
+            return R(minOr(this, rhs), maxOr(this, rhs));
+        }
+
+        R l = this.copy();
+        R r = rhs.copy();
+
+        // both intervals span [-1,0]
+        if ((l.imin.value ^ l.imax.value) < 0 && (r.imin.value ^ r.imax.value) < 0)
+        {
+            // cannot be smaller than either l.min or r.min, set the other one to 0
+            SignExtendedNumber min = l.imin.value < r.imin.value ? l.imin : r.imin;
+
+            // only negative numbers for minimum
+            l.imin.value = 0;
+            r.imin.value = 0;
+
+            return R(min, maxOr(l, r));
+        }
+        else
+        {
+            // only one interval spans [-1,0]
+            if ((imin.value ^ imax.value) < 0) swap(l, r); // r spans [-1,0]
+
+            auto minOrNeg = minOr(l, R(r.imin, SignExtendedNumber(-1)));
+            auto minOrPos = minOr(l, R(SignExtendedNumber(0), r.imax));
+            auto maxOrNeg = maxOr(l, R(r.imin, SignExtendedNumber(-1)));
+            auto maxOrPos = maxOr(l, R(SignExtendedNumber(0), r.imax));
+
+            auto min = minOrNeg < minOrPos ? minOrNeg : minOrPos;
+            auto max = maxOrNeg > maxOrNeg ? maxOrNeg : maxOrPos;
+
+            return R(min, max);
+        }
+    }
+
+    R opBinary(string op:"^")(R rhs) const
+    {
+        return this & ~rhs | ~this & rhs;
+    }
+
+    R opBinary(string op:"+")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+    R opBinary(string op:"-")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+    R opBinary(string op:"*")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+    R opBinary(string op:"/")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+    R opBinary(string op:"%")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+    R opBinary(string op:"<<")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+    R opBinary(string op:">>")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+    R opBinary(string op:"^^")(R rhs)
+    {
+        // Not yet implemented
+        assert(0);
+    }
+
+private:
+    // Credits to Timon Gehr maxOr, minOr, maxAnd, minAnd
+    // https://github.com/tgehr/d-compiler/blob/master/vrange.d
+    static SignExtendedNumber maxOr(R lhs, R rhs){
```

Comment:
Formatting issues - brace should be on its own line and the function body should be indented.